[#11_1] Product domain

dding94 opened the pull request.

dding94 committed:
feat: Product 클래스 생성

임시로 만들어 둠, 맞게 만들었는지 잘 모름

DJ-archive commented:
빠른 구현을 위해 상품 등록, 관리 기능 코드를 작성하고 있었는데 '도메인 주도 개발 시작하기' 등의 책을 읽고 관련 자료들을 찾아보니 요구사항과 도메인을 이해하고 정리하는 것이 우선일 것 같습니다. 요구사항을 제대로 정하지 않게 되면 변경하거나 다시 만들어야할 코드와 설계가 많아지기 때문입니다. 명구님과 요구사항 정리 후 도메인과 데이터 모델링 초안을 어느정도 구성한 뒤에 코드를 작성하겠습니다..

dding94 committed:
feat: Product 도메인 생성

dding94 committed:
refactor: 생성자 구현

dding94 committed:
feat: ProductRepository 인터페이스 구현

dding94 commented on src/main/java/com/flab/livecommerce/domain/product/Product.java:
```diff
@@ -0,0 +1,34 @@
+    //옵션 그룹
+    private List<OptionGroup> optionGroups = new ArrayList<>();
+
+    public Product(
+        String name, String description, int price, int salesPrice,
+        int stockQuantity, List<OptionGroup> optionGroups
+    ) {
+        this.name = name;
+        this.description = description;
+        this.price = price;
+        this.salesPrice = salesPrice;
+        this.stockQuantity = stockQuantity;
+        this.optionGroups.addAll(optionGroups);
+    }
```
파라미터가 많을 경우에 하나의 객체(값 객체?)로 분리해볼 수 있을 것 같다는 생각이 들었는데..
클린코드에서 파라미터를 최소화 하라는 내용을 본적이 있어 생성자에도 동일하게 적용되는 내용일까요?
그래서 분리를 해도 괜찮을지는 확신이 안서는데 책임에 따라 분리하는게 일반적인 방법이 될까요?
슬랙에서 책임얘기 하면서 상관없다고 했는데 다시생각해보니 잘못됐다고 생각합니다.
Money 객체같은 경우에도 그렇게 분리를 하기 때문에..

2eve2 replied:
현재는 이대로 하고, 추후 비즈니스 로직이 많아져서 따로 객체를 분리하는 것이 좋다고 느껴질 때 분리해도 좋을 것 같습니다. :)

f-lab-bright replied:
그룹을 만들기 적절한 녀석이 있다면 그렇게 해보셔도 됩니다.
하지만 지금은 ddd 개념에서의 valueObject로 만드는것 외에는 여러 파라미터를 하나로 만들만한건 그다지 없어보이긴 하네요 :)
파라미터가 많아지는건 생성자에서는 어느정도 감안하셔도 됩니다만, 이게 복잡해지면 생성할 때 코드가 지저분해지거나, 필드가 추가될 때 어려움이 있을 수 있어서, 빌더 패턴으로 해결하는게 더 바람직 할 수 있습니다 :)

2eve2 committed:
fix: 상품 등록

- command 변환 추후 구현
- Repository Adaptor 추후 구현

2eve2 committed:
feat: 상품 등록시 이름, 모델명 중복 확인

dding94 committed:
refactor: 클래스 이름변경 및 폴더위치 변경

dding94 committed:
feat: 공통 예외처리 기능 추가

dding94 committed:
refactor: Api 명칭, 코드 네임 리팩토링

dding94 committed:
docs: api 테스트 파일 추가

dding94 committed:
chore: Product -> item 명칭 변경

이유: 단어길이를 줄이기 위함

dding94 committed:
docs: item-api 형식 수정

dding94 committed:
refactor: 동일한 bean 이름 수정 및 사용하지 않는 설정 삭제

MybatisConfig 는 설정안해도 이미 자동으로 yml 파일에서 셋팅이 됨.

dding94 committed:
refactor: 상품 도메인 리팩토링, 레이어 의존방향 맞도록 수정

도메인 int -> Integer 수정, 생성자에서 옵션그룹 제외 (임시)
애플리케이션 레이어가 프레젠테이션 레이어 의존하는 코드 삭제, 추가로 이름도 명확하게 바꿨습니다.

dding94 committed:
refactor: NotBlank -> NotNull 수정

NotBlank 는 String 만 사용가능하기 때문에 예외가 발생함

dding94 committed:
feat: 상품 저장 기능 구현

인메모리 방식으로 구현

dding94 committed:
refactor: 상품 등록 컨트롤러 수정

검증로직이 해당부분에 있어야 하는가 생각이 들어 일단은 빼뒀습니다.

dding94 committed:
refactor: 인라인 리팩토링

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/item/config/ItemProcessorConfig.java:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class ItemProcessorConfig {
```
ProcessorConfig 라고 설정하면 유저에 있는 ProcessorConfig 와 충돌이 일어나 BeanDefinitionStoreException 예외를 만들어 이름 변경했습니다.

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/config/MybatisConfig.java:
```diff
@@ -1,22 +0,0 @@
-import org.springframework.context.annotation.Configuration;
-import org.springframework.context.annotation.Primary;
-
-@Configuration
-public class MybatisConfig {
-
-    @Bean
-    @Primary
-    @ConfigurationProperties(prefix = "spring.datasource")
-    public HikariDataSource hikariDataSource() {
-        return DataSourceBuilder
-            .create()
-            .type(HikariDataSource.class)
-            .build();
-    }
-}
```
해당 부분 pr#10 에 리뷰남긴것처럼 yml에서 설정했기 때문에 따로 설정할 필요가 없습니다.
이 부분을 빈으로 등록해버리면 오류가 발생합니다.
그래서 아얘 삭제를 했는데 제가 잘못 인지하거나 사용하신 이유가 있으시면 코멘트 부탁드리겠습니다!

2eve2 replied:
yml 설정때문에 해당 파트 지워진거 인지하고 있었는데 다른 브랜치에서 이 브랜치로 넘어올 때 달라진 부분 적용이 제대로 안되었던 것 같습니다. 주의하겠습니다.

dding94 commented on src/main/java/com/flab/livecommerce/presentation/item/ItemController.java:
```diff
@@ -0,0 +1,27 @@
+
+    @PostMapping
+    public CommonApiResponse registerItem(@RequestBody @Valid RegisterItemRequest request) {
+        itemManager.register(request.toCommand());
+        return CommonApiResponse.success(null);
```
지난 커밋에 이부분에 검증메서드를 넣으셨는데 해당 위치보다는 애플리케이션 레이어 쪽에 위치하도록 하는 방법은 어떻게 생각하시나요?

2eve2 replied:
컨트롤러 단의 메서드에서 비즈니스 로직을 노출시키는 것이 좋은 것 같아 검증 메서드를 노출시켰는데, 검증도 상품 등록시 일어나는 일이니 애플리케이션에 합쳐도 될 것 같습니다.

2eve2 replied:
검증 로직을 애플리케이션 레이어에 위치하도록만 하시고 상품 등록할 때 실제로 검증하도록 메서드 호출하지 않으셨네요. 메서드 호출하게끔 바꾸겠습니다.

dding94 replied:
네 검증은 일단 다 빼두고 구현만 해둔 상태로 만든겁니다. 알겠습니다

dding94 commented on src/main/java/com/flab/livecommerce/presentation/item/request/RegisterItemRequest.java:
```diff
@@ -0,0 +1,52 @@
+    @NotBlank(message = "상품명을 작성하세요.")
+    private String name;
+
+    @NotNull(message = "상품 원가를 작성하세요.")
+    private Integer price;
```
커밋으로 이유 써드리긴 했는데 NotBlank 는 String 만 적용이 됩니다.
그래서 에러가 발생해서 이부분은 일단 NotNull 로 수정을 했습니다! 
숫자관련해서 애노테이션이 많은데 요구사항에 맞춰서 다른 애노테이션을 적용하면 좋을 것 같습니다.

2eve2 replied:
좋은 지적 감사합니다. 검증 관련 어노테이션은 간단히 알고 있었던 지식을 바탕으로 사용했었는데 같은 기능을 하더라도 타입에 따라 어노테이션 사용이 어떻게 달라지는지 확인해보면서 쓰겠습니다!

dding94 committed:
Merge remote-tracking branch 'origin/main' into feature/11-product

# Conflicts:
#	src/main/java/com/flab/livecommerce/infrastructure/config/MybatisConfig.java

dding94 committed:
chore: 병합하면서 필요없는 문서 포함되어 삭제

2eve2 committed:
fix: 도커 env 파일 수정

- application.yml의 db username에 맞춰 도커 설정 파일 수정 및 db connection 에러 해결

2eve2 committed:
refactor: 상품 검증 추가, api response data

2eve2 reviewed (approved):
논의했던 대로 옵션과 상품 crud 나머지 작업은 새로운 pr에서 하겠습니다.
